### PR TITLE
Tightly coupled market-commons with prediction markets.

### LINKED ISSUE #805: [PM][Swaps] Tightly couple market-commons pallet
The market-commons pallet was spawned in the past to avoid circular dependencies between the prediction markets and swaps pallet. Both pallets strictly depend on it, and when executing migrations, often they need access to its storage. This is a sign that market-commons should be tightly coupled to those pallets.

### PR DESCRIPTION
**Pull request by vivekvpandya**
Fixes #805

### CODE REVIEW

**Review by maltekliemann — COMMENTED** [review]
Double-booking the implementation of `MarketCommonsPalletApi` into the implementation of the `zrml_market_commons::Pallet::<T>` seems strange. Is there a particular reason why you did that, instead of replacing every `zrml_market_commons::Pallet::<T>` with `<zrml_market_commons::Pallet::<T> as MarketCommonsPalletApi>`?

**Comment by vivekvpandya** [comment]
> Double-booking the implementation of `MarketCommonsPalletApi` into the implementation of the `zrml_market_commons::Pallet::<T>` seems strange. Is there a particular reason why you did that, instead of replacing every `zrml_market_commons::Pallet::<T>` with `<zrml_market_commons::Pallet::<T> as MarketCommonsPalletApi>`?

Frankly I did not get this idea when editing, I though for loose coupling we use MarketCommonsPalletApi and for tight coupling we directly use functions from pallet and don't need to import MarketCommonsPalletApi.

**Comment by Chralt98** [comment]
Starting my review now..

**Comment by maltekliemann** [comment]
> > Double-booking the implementation of `MarketCommonsPalletApi` into the implementation of the `zrml_market_commons::Pallet::<T>` seems strange. Is there a particular reason why you did that, instead of replacing every `zrml_market_commons::Pallet::<T>` with `<zrml_market_commons::Pallet::<T> as MarketCommonsPalletApi>`?
> 
> Frankly I did not get this idea when editing, I though for loose coupling we use MarketCommonsPalletApi and for tight coupling we directly use functions from pallet and don't need to import MarketCommonsPalletApi.

It's a matter of dependency. With loose coupling, you can _only_ access the config parameter `MarketCommons`, which implements `MarketCommonsPalletApi`. This allows you to use _any_ struct which implements that particular trait. With tight coupling, that config parameter is gone and we have to use the particular `zrml_market_commons` pallet, which happens to implement `MarketCommonsPalletApi`. In particular, the user has no choice in the matter. Before this change, they could have used their own "market manager". Now they _have_ to use ours.

But since the MDMs are still loosely coupled, I think we should move `MarketCommonsPalletApi` into the `primitives`. Is there any argument against this?

**Comment by Chralt98** [comment]
You like to have the tight coupling of `zrml_swaps` in another PR?

**Comment by vivekvpandya** [comment]
> You like to have the tight coupling of `zrml_swaps` in another PR?

yes if it is really required, as per discussion in one of the meeting we feel that swaps should not be tightly coupled with market-commons.

File: zrml/market-commons/src/lib.rs [hunk 1]
```diff
@@ -152,10 +158,7 @@ mod pallet {
             <Markets<T>>::iter()
         }
 
-        fn market(
-            market_id: &Self::MarketId,
-        ) -> Result<Market<Self::AccountId, Self::BlockNumber, Self::Moment>, DispatchError>
-        {
+        fn market(market_id: &Self::MarketId) -> Result<MarketOf<T>, DispatchError> {
```
Comment (Chralt98) [inline zrml/market-commons/src/lib.rs:161]: You can use this cool approach [here](https://github.com/zeitgeistpm/zeitgeist/blob/d6a3b8a5e1d056127c51f00f2f02be3409ebe742/zrml/market-commons/src/lib.rs#L181), [here](https://github.com/zeitgeistpm/zeitgeist/blob/d6a3b8a5e1d056127c51f00f2f02be3409ebe742/zrml/market-commons/src/lib.rs#L168), [here](https://github.com/zeitgeistpm/zeitgeist/blob/d6a3b8a5e1d056127c51f00f2f02be3409ebe742/zrml/market-commons/src/lib.rs#L238) and [here](https://github.com/zeitgeistpm/zeitgeist/blob/d6a3b8a5e1d056127c51f00f2f02be3409ebe742/zrml/market-commons/src/lib.rs#L156).

File: zrml/prediction-markets/src/lib.rs [hunk 2]
```diff
@@ -122,10 +121,11 @@ mod pallet {
             // TODO(#618): Not implemented for Rikiddo!
             T::DestroyOrigin::ensure_origin(origin)?;
 
-            let market = T::MarketCommons::market(&market_id)?;
+            let market =
+                <zrml_market_commons::Pallet<T> as MarketCommonsPalletApi>::market(&market_id)?;
```
Comment (Chralt98) [inline zrml/prediction-markets/src/lib.rs:125]: You can globally replace `<zrml_market_commons::Pallet<T> as MarketCommonsPalletApi>` by `<zrml_market_commons::Pallet<T>>` to shorten things up here.

```suggestion
            let market = <zrml_market_commons::Pallet<T>>::market(&market_id)?;
```

**Review by Chralt98 — CHANGES_REQUESTED** [review]
I like the idea of the tightly couple.

File: zrml/prediction-markets/src/lib.rs (hunk 2 above)
Comment (vivekvpandya) [reply zrml/prediction-markets/src/lib.rs:125]: I wonder why clippy don't suggest that :thinking:

**Comment by vivekvpandya** [comment]
> But since the MDMs are still loosely coupled, I think we should move MarketCommonsPalletApi into the primitives. Is there any argument against this?

so what is the motivation here? Just that above trait is imported from primitives and not from market_commons? I mean only implementer of that trait is MarketCommons so I am fine with any approach.

File: zrml/prediction-markets/src/lib.rs (hunk 2 above)
Comment (maltekliemann) [reply zrml/prediction-markets/src/lib.rs:125]: I wonder why I suggested it. My bad, of course we don't need the fully qualified syntax if there's no collisions between the function names.

**Comment by maltekliemann** [comment]
> > But since the MDMs are still loosely coupled, I think we should move MarketCommonsPalletApi into the primitives. Is there any argument against this?
> 
> so what is the motivation here? Just that above trait is imported from primitives and not from market_commons? I mean only implementer of that trait is MarketCommons so I am fine with any approach.

Because the MDMs and `swaps` are still loosely coupled, people could potentially use them with some other struct that implements `MarketCommonsPalletApi`. If they wanted to do that, they would have to import `zrml_market_commons` to get the trait, which is an unnecessary import. Instead, they should be able to import the trait from the `primitives`.

File: zrml/prediction-markets/src/benchmarks.rs [hunk 3]
```diff
@@ -98,7 +98,7 @@ fn create_market_common<T: Config>(
         scoring_rule,
     }
     .dispatch_bypass_filter(RawOrigin::Signed(caller.clone()).into())?;
-    let market_id = T::MarketCommons::latest_market_id()?;
+    let market_id = <zrml_market_commons::Pallet<T> as MarketCommonsPalletApi>::latest_market_id()?;
```
Comment (maltekliemann) [inline zrml/prediction-markets/src/benchmarks.rs:101]: This is inconsistent use of syntax. You use Chris' suggestion in `prediction-markets`, but my (bad) suggestion here.

File: zrml/prediction-markets/src/lib.rs [hunk 4]
```diff
@@ -2613,6 +2620,10 @@ mod pallet {
             T::PalletId::get().0
         }
 
+        fn market_account(market_id: MarketIdOf<T>) -> T::AccountId {
+            T::PalletId::get().into_sub_account_truncating(market_id.saturated_into::<u128>())
+        }
+
```
Comment (maltekliemann) [inline zrml/prediction-markets/src/lib.rs:2626]: Why did you add this again?

```suggestion
```

File: zrml/prediction-markets/src/lib.rs [hunk 5]
```diff
@@ -2709,12 +2720,12 @@ mod pallet {
...
             outcome_report: &OutcomeReport,
         ) -> Result<Weight, DispatchError> {
-            let pool_id = if let Ok(el) = T::MarketCommons::market_pool(market_id) {
+            let pool_id = if let Ok(el) = <zrml_market_commons::Pallet<T>>::market_pool(market_id) {
                 el
             } else {
                 return Ok(T::DbWeight::get().reads(1));
             };
-            let market_account = T::MarketCommons::market_account(*market_id);
+            let market_account = Self::market_account(*market_id);
```
Comment (maltekliemann) [inline zrml/prediction-markets/src/lib.rs:2728]:
```suggestion
            let market_account = <zrml_market_commons::Pallet<T>>::market_account(*market_id);
```

**Review by maltekliemann — CHANGES_REQUESTED** [review]

File: zrml/prediction-markets/src/benchmarks.rs (hunk 3 above)
Comment (vivekvpandya) [reply zrml/prediction-markets/src/benchmarks.rs:101]: oh missed this file.

File: zrml/prediction-markets/src/lib.rs (hunk 4 above)
Comment (vivekvpandya) [reply zrml/prediction-markets/src/lib.rs:2626]: I think that now we have tight coupling we may perhaps able to remove `PredictionMarketsPalletId` from market_commons Config. however if we want to keep `market_account()` in the trait then I accept your suggested edit.

Comment (vivekvpandya) [reply zrml/prediction-markets/src/lib.rs:2626]: but I think we can't get rid of `PredictionMarketsPalletId` as we can't access prediction_market config in matket_commons.

**Review by Chralt98 — APPROVED** [review]
LGTM

**Review by maltekliemann — APPROVED** [review]
:thumbsup:
